datafusion/optimizer/src/eliminate_group_by_constant.rs

```diff
@@ -0,0 +1,318 @@
+
+use datafusion_common::tree_node::Transformed;
+use datafusion_common::{internal_err, Result};
+use datafusion_expr::{Aggregate, Expr, LogicalPlan, LogicalPlanBuilder, Volatility};
+
+/// Optimizer rule that removes constant expressions from `GROUP BY` clause
+/// and places additional projection on top of aggregation, to preserve
+/// original schema
+#[derive(Default)]
+pub struct EliminateGroupByConstant {}
+
+impl EliminateGroupByConstant {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for EliminateGroupByConstant {
+    fn supports_rewrite(&self) -> bool {
+        true
+    }
+
+    fn rewrite(
+        &self,
+        plan: LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Transformed<LogicalPlan>> {
+        match plan {
+            LogicalPlan::Aggregate(aggregate) => {
+                let (const_group_expr, nonconst_group_expr): (Vec<_>, Vec<_>) = aggregate
+                    .group_expr
+                    .iter()
+                    .partition(|expr| is_constant_expression(expr));
+
+                // If no constant expressions found (nothing to optimize) or
+                // constant expression is the only expression in aggregate,
+                // optimization is skipped
+                if const_group_expr.is_empty()
+                    || (!const_group_expr.is_empty()
+                        && nonconst_group_expr.is_empty()
+                        && aggregate.aggr_expr.is_empty())
+                {
+                    return Ok(Transformed::no(LogicalPlan::Aggregate(aggregate)));
+                }
+
+                let simplified_aggregate = LogicalPlan::Aggregate(Aggregate::try_new(
+                    aggregate.input,
+                    nonconst_group_expr.into_iter().cloned().collect(),
+                    aggregate.aggr_expr.clone(),
+                )?);
+
+                let projection_expr =
+                    aggregate.group_expr.into_iter().chain(aggregate.aggr_expr);
+
+                let projection = LogicalPlanBuilder::from(simplified_aggregate)
+                    .project(projection_expr)?
+                    .build()?;
+
+                Ok(Transformed::yes(projection))
+            }
+            _ => Ok(Transformed::no(plan)),
+        }
+    }
+
+    fn try_optimize(
+        &self,
+        _plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        internal_err!("Should have called EliminateGroupByConstant::rewrite")
+    }
+
+    fn name(&self) -> &str {
+        "eliminate_group_by_constant"
+    }
+
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        Some(ApplyOrder::BottomUp)
+    }
+}
+
+/// Checks if expression is constant, and can be eliminated from group by.
+///
+/// Intended to be used only within this rule, helper function, which heavily
+/// reiles on `SimplifyExpressions` result.
+fn is_constant_expression(expr: &Expr) -> bool {
```

Comment:
I think you could use `TreeNode::Exists` here: https://github.com/apache/datafusion/blob/045e8fcaeac2f2c29afe5017e5efe5a3a2560080/datafusion/common/src/tree_node.rs#L408

Which handles all types of expressions and would be general

Something like this, perhaps (untested)

```rust
fn is_constant_expression(expr: &Expr) -> bool {
  !expr.exsts(|expr| {
    match expr { 
      Expr::Column(_) => true,
      Expr::ScalarFunction(e) if matches!(
                e.func.signature().volatility,
                Volatility::Volatile
      ) => true,
      _ => false
 })
```

Reply:
Thank you for the hint. I've checked id out, and seems like there is no TreeNode API method for this case, since it requires all arguments of scalar function to be literals, and `exists` will return on first match.